### ISSUE TITLE
Invalid cast not caught in compiler or runtime leading to dotnet.exe access violation.

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#42909]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/967375/invalid-cast-not-caught-in-compiler-or-runtime-lea.html)._

---
Both the C# compiler and Visual Studio Intellisense show an obviously invalid cast as valid.  This is leading to a dotnet.exe access violation when executed.  The code necessary to reproduce is in the following github repo.  The offending class is commented as such in Try.cs in BadCastAccessViolation:

https://github.com/slowsigma/FoundErrors

The class UnitTest1.cs in BadCastAccessViolation.Tests contains a single test where you can reproduce access violation.

public abstract class Try<T>
    {
        public static explicit operator Try<T>(Func<T> value) => (Pending<T>)value;

// VS calls this cast redundant, but there is no defined operator (explicit or implicit).
        //public static implicit operator Try<T>(Suspect<T> value) => (Resolved<T>)value;
        // So, if we indulge Visual Studio and take the suggestion to remove the "redundant" cast,
        // this statement leads un-catchable access violation
        public static implicit operator Try<T>(Suspect<T> value) => value;

public abstract Suspect<T> Run();
    }

---
#### Visual Studio Feedback System on 3/29/2020, 11:53 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by CyrusNajmabadi** [comment]
@sharwell combined repro.  The code does seem to be a bit.. strange:

```c#
class C
{
    static void Main(string[] args)
    {
        var c = new Confirmed<int>(0);
        var t = (Try<int>)c;
    }
}

namespace Nysa.Logics
{

    public abstract class Suspect<T>
    {
    }

    public sealed class Confirmed<T> : Suspect<T>
    {
        public static implicit operator T(Confirmed<T> confirmed) => confirmed.Value;

        public T Value { get; }
        public Confirmed(T value) { this.Value = value; }
        public override String ToString() => this.Value.ToString();
    }


    public sealed class Failed<T> : Suspect<T>
    {
        public static explicit operator Failed<T>(Exception value) => new Failed<T>(value);
        public static implicit operator Exception(Failed<T> failed) => failed.Value;

        public Exception Value { get; }
        public Failed(Exception value) { this.Value = value; }
        public override String ToString() => this.Value.Message;
    }

    public sealed class Pending<T> : Try<T>
    {
        public static explicit operator Pending<T>(Func<T> value) => new Pending<T>(value);

        private Func<T> Value { get; }

        private Pending(Func<T> value) { this.Value = value; }

        public override Suspect<T> Run()
        {
            try { return this.Value().Confirmed(); }
            catch (Exception except) { return except.Failed<T>(); }
        }
    }

    
    public static class SuspectExtensions
    {
        public static Suspect<T> Confirmed<T>(this T @this) => new Confirmed<T>(@this);
        public static Suspect<T> Failed<T>(this Exception @this) => new Failed<T>(@this);
    }

    public sealed class Resolved<T> : Try<T>
    {
        private Suspect<T> Value { get; }

        private Resolved(Suspect<T> value) { this.Value = value; }

        public override Suspect<T> Run() => this.Value;
    }


    public abstract class Try<T>
    {
        public static explicit operator Try<T>(Func<T> value) => (Pending<T>)value;

        // VS calls this cast redundant, but there is no defined operator (explicit or implicit).
        public static implicit operator Try<T>(Suspect<T> value) => (Resolved<T>)value;

        public abstract Suspect<T> Run();
    }
}
```

The cast is redundant here.  Removing is has no change in behavior.  The code *both* before and *after* stack-overflows. That's because the only conversion C# can ind to handle `(Resolved<T>)value` is the implicit operator that that code is currently in.  so that cast calls `implicit operator Try<T>(Suspect<T> value)`, which calls itself, which calls itself.

**Comment by slowsigma** [comment]
Not sure the resolution is justified.  The error that occurred was not a stack overflow it was an access violation that the runtime did not catch.  The original explicit cast "... => (Resolved<T>)value;" should have shown as an error because it didn't exist.  Also, is there no compiler warning on implicit operators that call themselves?

**Comment by CyrusNajmabadi** [comment]
> The original explicit cast "... => (Resolved)value;" should have shown as an error because it didn't exist.

No the original cast does exist.  From the language's perspective you have defined that cast here:

```
public static implicit operator Try<T>(Suspect<T> value)
```

That's why teh code compiles without error even with your original cast in it.

**Comment by CyrusNajmabadi** [comment]
> Also, is there no compiler warning on implicit operators that call themselves?

No, there is not.  It's a potential space for an analyzer.

**Comment by CyrusNajmabadi** [comment]
> Not sure the resolution is justified. The error that occurred was not a stack overflow it was an access violation that the runtime did not catch

I do not repro that.  I get a stack overflow in all cases.  If you get an AV, please file an issue over at dotnet/runtime.  This is the repo for the compiler, and there seems to be no issue there.

**Comment by slowsigma** [comment]
Got it.  I was stuck not seeing the implicit operation I was defining.  I only went down the rabbit hole on this unfinished code because of what VS was showing and not showing.  I probably would have just continued finishing out the code if not for the uncaught access violation.  Not sure how you repro different unless it has something to do with xUnit.
